### DISCUSSION TITLE
Web Deploy error "file not found"

### DISCUSSION THREAD

**Discussion by qtec-tm** [serenity-is/Serenity#6973] [Help] [answered]
Hello,

I have a problem deploying my Serene project to an IIS webserver using Web Deploy.

Build runs fine, but then these messages appear:
```
[...]
node tsbuild

  wwwroot\esm\_chunks\chunk-T4JWLVJO.js                              210.2kb
  wwwroot\esm\_chunks\chunk-PMD3JE7K.js                                6.6kb
  wwwroot\esm\_chunks\chunk-WQWXCWHB.js                                5.1kb
  wwwroot\esm\_chunks\chunk-DWM5OA25.js                                5.0kb
  wwwroot\esm\Modules\Administration\Translation\TranslationPage.js    4.1kb
  ...and 71 more output files...

Done in 307ms
esbuild clean: deleting extra file _chunks\chunk-NNGTOG2N.js
esbuild clean: deleting extra file _chunks\chunk-NNGTOG2N.js.map
C:\Program Files\dotnet\sdk\7.0.110\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(593,5): Error MSB3030: Could not copy the file "C:\dev\hsw\XXXXXXXXX\XXXXXXXXX\XXXXXXXXX.Web\wwwroot\esm\_chunks\chunk-NNGTOG2N.js" because it was not found.
C:\Program Files\dotnet\sdk\7.0.110\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(593,5): Error MSB3030: Could not copy the file "C:\dev\hsw\XXXXXXXXX\XXXXXXXXX\XXXXXXXXX.Web\wwwroot\esm\_chunks\chunk-NNGTOG2N.js.map" because it was not found.
```

Seems like `esbuild clean` deletes files, which should be copied to the target.
I have no clue where to start. For 6 years now it always worked, after updating to the recent version and running through various update migrations publishing using web deploy has stopped working.

Can someone give me a hint?

Thanks in advance
Thomas

**Comment by qtec-tm** [comment]
Update:

Removing the target `CompileTSC` from .csproj file helps. But it is necessary to rebuild before publishing and in addition errors like described [here](https://github.com/orgs/serenity-is/discussions/6775) may appear on the target server. Therefore it is necessary to re-publish the solution until it works.

But that's just a workaround, not the solution.

**Comment by volkanceylan** [comment, marked as answer]
It is an issue with ASP.NET Core static web assets, and a chicken-egg situation. Your project must be built, and esbuild runned successfully before publishing. Otherwise, some files which were available before the build starts, might not be there before the build ends (as esbuild is run) so static web assets calculation is getting confused. They are not static web assets actually, but as it thinks everything under wwwroot are static web assets, it raises such errors. You may ignore it for now, or build first, run tsbuild, then publish.

**Comment by qtec-tm** [comment]
Thank you for your answer. That means every time before I publish, I have to remove the target `CompileTSC` in the `.csproj` file because Visual Studio automatically builds when I publish.

**Comment by YoselYohan** [comment]
This might not be related to your problem. Recently with updates I found this issue that was intermittent when compiling/publishing. I installed .NET SDK 8.0.404 (I had 7 and 9 installed) and placed a file "global.json" in my project directory and it worked!!!.

this is the global.json content
`{
  "sdk": {
    "version": "8.0.404",
    "rollForward": "latestPatch"
  }
}`

you can find info [here](https://learn.microsoft.com/en-us/dotnet/core/versions/selection)
Done in 1448ms
esbuild clean: deleting extra file _chunks\chunk-NJP4PYJA.js
esbuild clean: deleting extra file _chunks\chunk-NJP4PYJA.js.map
C:\Program Files\dotnet\sdk\9.0.101\Sdks\Microsoft.NET.Sdk.StaticWebAssets\targets\Microsoft.NET.Sdk.StaticWebAssets.targets(662,5): Error : System.InvalidOperationException: No file exists for the asset at either location 'wwwroot\esm\_chunks\chunk-NJP4PYJA.js' or 'wwwroot\esm\_chunks\chunk-NJP4PYJA.js'.
   at Microsoft.AspNetCore.StaticWebAssets.Tasks.StaticWebAsset.ComputeFingerprintAndIntegrity(String identity, String originalItemSpec)
   at Microsoft.AspNetCore.StaticWebAssets.Tasks.DefineStaticWebAssets.Execute()

**Reply by qtec-tm** [reply]
Hello,
thank you for your answer.
In the meantime I made a new .NET 8 project and migrated the whole project to .NET 8 and TS modules. I also searched for circular dependendies using npm module madge. This fixed this issue and another issue mentioned [here](https://github.com/orgs/serenity-is/discussions/6775).
Making these changes took me over a week but now the problems are gone.
